### PR TITLE
2 bot behavior types

### PR DESCRIPTION
**Pull request by ghost**
1. Bot that runs away when sees the player.
2. Bot that chases the player when see him.
#5

### CODE REVIEW

File: Source/Prot/AI/BTRunPointSelect.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	ABotAIC* AIController = Cast<ABotAIC>(OwnerComp.GetAIOwner());
+	if (AIController)
+	{
+		UBlackboardComponent* BlackboardComponent = AIController->GetBlackboardComponent();
+		AActor* Player = Cast<AActor>(BlackboardComponent->GetValueAsObject("TargetPlayer"));
+		AActor* Bot = Cast<AActor>(BlackboardComponent->GetValueAsObject("SelfActor"));
+		if (Player && Bot)
+		{
+			float angle = FMath::RandRange(-135.f, 135.f);
+			FVector direction = UKismetMathLibrary::GetDirectionUnitVector(Player->GetActorLocation(), Bot->GetActorLocation());
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTRunPointSelect.cpp]: Don't you remember naming convention? Variable names start with capitals.

File: Source/Prot/AI/BTRunPointSelect.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+class PROT_API UBTRunPointSelect : public UBTTaskNode
+{
+	GENERATED_BODY()
+	
+	virtual EBTNodeResult::Type ExecuteTask(UBehaviorTreeComponent& OwnerComp, uint8* NodeMemory) override;
+
+public:
+	// The distance for Bot to run from Player
+	UPROPERTY(Category = Node, EditAnywhere, meta = (ClampMin = "0.0", UIMin = "0.0"))
+	float RunDistance;
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTRunPointSelect.h:22]: Do you mean something like `KeepAwayRange`?

File: Source/Prot/AI/BotAIC.cpp [hunk 3]
```diff
@@ -28,6 +28,7 @@ void ABotAIC::Possess(APawn* Pawn)
 		{
 			BlackboardComponent->InitializeBlackboard(*(Bot->BehaviorTree->BlackboardAsset));
 			BlackboardComponent->SetValueAsInt("PointIndex", 0);
+			BlackboardComponent->SetValueAsObject("SelfActor", Bot);
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp:31]: Can't we access `ABot` from the blackboard?

**Review by UnoYakshi — CHANGES_REQUESTED** [review]

File: Source/Prot/AI/BotAIC.cpp (hunk 3 above)
Comment (ghost) [reply Source/Prot/AI/BotAIC.cpp:31]: Bot is used for blackboard initializing at 29 line

File: Source/Prot/AI/BTRunPointSelect.h [hunk 4]
```diff
@@ -19,5 +19,5 @@ class PROT_API UBTRunPointSelect : public UBTTaskNode
 public:
 	// The distance for Bot to run from Player
 	UPROPERTY(Category = Node, EditAnywhere, meta = (ClampMin = "0.0", UIMin = "0.0"))
-	float RunDistance;
+	float KeepAwayRange;
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTRunPointSelect.h]: It was a question... Because it's unclear from the comment.

Comment (ghost) [reply Source/Prot/AI/BTRunPointSelect.h]: What is unclear exactly?

**Review by andrew-it — APPROVED** [review]

**Comment by ghost** [comment]
### Coward Bot
Coward Bot
### Agressive Bot
Agressive Bot

File: Source/Prot/AI/BTRunPointSelect.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+#include "Math/UnrealMathUtility.h"
+#include "Kismet/KismetMathLibrary.h"
+
+EBTNodeResult::Type UBTRunPointSelect::ExecuteTask(UBehaviorTreeComponent & OwnerComp, uint8 * NodeMemory)
+{
+	ABotAIC* AIController = Cast<ABotAIC>(OwnerComp.GetAIOwner());
+	if (AIController)
+	{
+		UBlackboardComponent* BlackboardComponent = AIController->GetBlackboardComponent();
+		AActor* Player = Cast<AActor>(BlackboardComponent->GetValueAsObject("TargetPlayer"));
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTRunPointSelect.cpp]: Check if `BlackboardComponent` isn't `nullptr`.

File: Source/Prot/AI/BTRunPointSelect.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+			FVector Direction = UKismetMathLibrary::GetDirectionUnitVector(Player->GetActorLocation(), Bot->GetActorLocation());
+			// Rotate direction so Bot will run away from Player not towards
+			Direction = Direction.RotateAngleAxis(Angle, FVector(0, 0, 1));
+			// Create a point where Bot will run from Player
+			BlackboardComponent->SetValueAsVector("RunLocation", Bot->GetActorLocation() + Direction * RunDistance);
+			return EBTNodeResult::Succeeded;
+		}
+	}
+	return EBTNodeResult::Failed;
+}
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTRunPointSelect.cpp]: Add a newline.

File: Source/Prot/AI/BotAIC.cpp (hunk 3 above)
Comment (UnoYakshi) [reply Source/Prot/AI/BotAIC.cpp:31]: Then what does `SelfActor` mean in this case?

**Review by UnoYakshi — CHANGES_REQUESTED** [review]

**Review by UnoYakshi — APPROVED** [review]

**Pull request merged**